### PR TITLE
Refactor Saved objects and add visualization embeddable

### PR DESCRIPTION
**Pull request by mengweieric**
### Description
This PR has two main changes:
1. add saved visualization embeddable
    - add `observability-visualization` saved object type
    - add `observability-visualization` embeddable type to show saved visualizations in core dashboards
    - saved object id are in the form of `type:uuid`, e.g. `observability-visualization:66cfdf30-d310-11ed-a37b-4350efe216f8`
1. refactor saved objects:
    - create clients for handling observability and OSD saved objects
    - update new saved visualization to use OSD saved objects client
    - update rest of event analytics, panels, notebooks to use saved objects clients based on object id
    - other refactor changes from Eric

The embeddable changes are only in the `public/embeddable`, `server/saved_objects` directory and the `plugin.ts` files. Rest are refactoring. Will add more tests in another PR.

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/types/observability_saved_object_attributes.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+export const OBSERVABILTY_SAVED_OBJECTS = [VISUALIZATION_SAVED_OBJECT] as const;
+export const SAVED_OBJECT_VERSION = 1;
+
+export interface VisualizationSavedObjectAttributes extends SavedObjectAttributes {
+  title: string;
+  description: string;
+  version: number;
+  createdTimeMs: number;
+  savedVisualization: SavedVisualization;
+}
```
Comment (ps48) [inline common/types/observability_saved_object_attributes.ts:19]: I see that we are using the public service for SavedObjects. The server service has more options for something like version: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/416eb508b3b22fe81231b726dcfce09e3aad9bfb/src/core/server/saved_objects/service/saved_objects_client.ts#L47-L71
Did we take a look at it?

File: common/utils/index.ts [hunk 2]
```diff
@@ -10,4 +10,5 @@ export {
   composeFinalQuery,
   removeBacktick,
 } from './query_utils';
-export { uiSettingsService } from './settings_service';
+
+export * from './core_services';
```
Comment (ps48) [inline common/utils/index.ts:14]: Should we use `*` here? Is linting allowing it?

File: common/utils/query_utils.ts [hunk 3]
```diff
@@ -65,7 +66,13 @@ export const preprocessQuery = ({
 
   finalQuery = `${tokens![1]}=${
     tokens![2]
-  } | where ${timeField} >= '${start}' and ${timeField} <= '${end}'${tokens![3]}`;
+  } | where ${timeField} >= '${start}' and ${timeField} <= '${end}'`;
+
+  if (whereClause) {
+    finalQuery += ` AND ${whereClause}`;
+  }
```
Comment (ps48) [inline common/utils/query_utils.ts:73]: nice.

File: common/types/observability_saved_object_attributes.ts (hunk 1 above)
Comment (joshuali925) [reply common/types/observability_saved_object_attributes.ts:19]: version can be passed as a create option in server but not in frontend. https://github.com/opensearch-project/opensearch-dashboards/blob/11b98ec05483269917c335fcb1900bf98da8cac6/src/core/public/saved_objects/saved_objects_client.ts#L54

If we want to use it then saved object client should be called in server? But this seems to be the version of a specific object that will change on each update, not sure if it would be useful for us

```
  /**
   * An opaque version number which changes on each successful write operation.
   * Can be used in conjunction with `overwrite` for implementing optimistic concurrency control.
   **/
  version?: string;
```

File: common/utils/index.ts (hunk 2 above)
Comment (joshuali925) [reply common/utils/index.ts:14]: lint didn't complain but I was going to remove functions in `core_services`. didn't find a way yet since embeddable is not rendered by observability plugin but also needs PPLService

File: common/utils/query_utils.ts (hunk 3 above)
Comment (joshuali925) [reply common/utils/query_utils.ts:73]: this is temporary and will change when we support DSL in PPL where clauses

File: common/types/observability_saved_object_attributes.ts (hunk 1 above)
Comment (ps48) [reply common/types/observability_saved_object_attributes.ts:19]: That isn't required then. 👍🏽

File: server/saved_objects/observability_saved_object.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    properties: {
+      title: {
+        type: 'text',
+      },
+      description: {
+        type: 'text',
+      },
+      version: { type: 'integer' },
+    },
+  },
```
Comment (ps48) [inline server/saved_objects/observability_saved_object.ts:41]: Are we finalizing on mappings or is that subject to change?

File: public/services/saved_objects/saved_object_loaders/ppl/index.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Comment (ps48) [inline public/services/saved_objects/saved_object_loaders/ppl/index.ts:4]: Is this file needed?

**Review by ps48 — APPROVED** [review]

File: public/components/notebooks/components/paragraph_components/paragraphs.tsx [hunk 6]
```diff
@@ -510,6 +511,7 @@ export const Paragraphs = forwardRef((props: ParagraphProps, ref) => {
     <>
       <EuiPanel>
         {renderParaHeader(!para.isVizualisation ? 'Code block' : 'Visualization', index)}
+        {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events */}
```
Comment (derek-ho) [inline public/components/notebooks/components/paragraph_components/paragraphs.tsx:514]: remove?

File: server/saved_objects/observability_saved_object.ts (hunk 4 above)
Comment (joshuali925) [reply server/saved_objects/observability_saved_object.ts:41]: this can change, will panels have mappings for specific fields?

File: public/services/saved_objects/saved_object_loaders/ppl/index.ts (hunk 5 above)
Comment (joshuali925) [reply public/services/saved_objects/saved_object_loaders/ppl/index.ts:4]: @mengweieric

**Review by joshuali925 — APPROVED** [review]
